OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs

```diff
@@ -0,0 +1,234 @@
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Support;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Manages AI MCVs.")]
+	public class McvManagerBotModuleInfo : ConditionalTraitInfo
+	{
+		[Desc("Actor types that are considered MCVs (deploy into base builders).")]
+		public readonly HashSet<string> McvTypes = new HashSet<string>();
+
+		[Desc("Actor types that are considered construction yards (base builders).")]
+		public readonly HashSet<string> ConstructionYardTypes = new HashSet<string>();
+
+		[Desc("Actor types that are able to produce MCVs.")]
+		public readonly HashSet<string> McvFactoryTypes = new HashSet<string>();
+
+		[Desc("Delay (in ticks) between looking for and giving out orders to new MCVs.")]
+		public readonly int ScanForNewMcvInterval = 20;
+
+		[Desc("Minimum distance in cells from center of the base when checking for MCV deployment location.")]
+		public readonly int MinBaseRadius = 2;
+
+		[Desc("Maximum distance in cells from center of the base when checking for MCV deployment location.",
+			"Only applies if RestrictMCVDeploymentFallbackToBase is enabled and there's at least one construction yard.")]
+		public readonly int MaxBaseRadius = 20;
+
+		[Desc("Should deployment of additional MCVs be restricted to MaxBaseRadius if explicit deploy locations are missing or occupied?")]
+		public readonly bool RestrictMCVDeploymentFallbackToBase = true;
+
+		public override object Create(ActorInitializer init) { return new McvManagerBotModule(init.Self, this); }
+	}
+
+	public class McvManagerBotModule : ConditionalTrait<McvManagerBotModuleInfo>, IBotTick, IBotPositionsUpdated
+	{
+		public CPos GetRandomBaseCenter()
+		{
+			var randomConstructionYard = world.Actors.Where(a => a.Owner == player &&
+				Info.ConstructionYardTypes.Contains(a.Info.Name))
+				.RandomOrDefault(world.LocalRandom);
+
+			return randomConstructionYard != null ? randomConstructionYard.Location : initialBaseCenter;
+		}
+
+		readonly World world;
+		readonly Player player;
+
+		readonly Predicate<Actor> unitCannotBeOrdered;
+
+		IBotPositionsUpdated[] notifyPositionsUpdated;
+		IBotRequestUnitProduction[] requestUnitProduction;
+
+		CPos initialBaseCenter;
+		int scanInterval;
+		int ticks;
+
+		// MCVs that the bot already knows about. Any MCV not on this list needs to be given an order.
+		List<Actor> activeMCVs = new List<Actor>();
+
+		public McvManagerBotModule(Actor self, McvManagerBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+
+			if (world.Type == WorldType.Editor)
```

Comment:
What's the purpose of this, and why don't we need it in the other module types?

Reply:
I don't remember, tbh. Presumably a left-over from outdated module plumbing that got copy-pasted around. I'll try removing it and see what happens (probably nothing).

Reply:
Removed it, didn't notice any regressions in the editor.